File: leetcode/longest_substring_without_repeating_characters.cpp
// Coder Agent: Longest Substring Without Repeating Characters (C++)
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> index;
        int maxLen = 0, left = 0;
        for (int right = 0; right < s.size(); ++right) {
            if (index.count(s[right]) && index[s[right]] >= left) {
                left = index[s[right]] + 1;
            }
            index[s[right]] = right;
            maxLen = max(maxLen, right - left + 1);
        }
        return maxLen;
    }
};

// Example usage
int main() {
    Solution sol;
    cout << sol.lengthOfLongestSubstring("abcabcbb") << endl; // Output: 3
    return 0;
}

File: leetcode/longest_substring_without_repeating_characters.java
// Coder Agent: Longest Substring Without Repeating Characters (Java)
import java.util.*;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> index = new HashMap<>();
        int maxLen = 0, left = 0;
        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (index.containsKey(c) && index.get(c) >= left) {
                left = index.get(c) + 1;
            }
            index.put(c, right);
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.lengthOfLongestSubstring("abcabcbb")); // Output: 3
    }
}
